# Задание **Lite**

1. Используя функцию с занятия (функция, которая будет добавлять новые данные в key-value storage)

    keys = [123]

    for k in keys:

      num_events = np.random.randint(0,5)

      new_val = np.random.uniform(low=0,high=50,size=num_events).tolist()

      kv_storage[k].extend(new_val)

2. Добавьте новые значения в key-value хранилище из Ultra-Light  для 1000 случайных ключей (из миллиона). 

3. Возьмите любой ключ, по которому были добавлены данные и выведите расчет метрик.

# Подключение библиотек

In [ ]:
import numpy as np
import random

# 1. Используя функцию с занятия

In [ ]:
kv_storage = {i:[] for i in range(1_000_000)}

In [ ]:
print("Длина словаря:", len(kv_storage))

Длина словаря: 1000000


In [ ]:
for key in kv_storage.keys():
  num_events = np.random.randint(0,320) # создаем случайное число
  new_val = np.random.uniform(low=0,high=50,size=num_events).tolist() # создаем массив размером в случайное число и преобразуем в список из numpy 
  kv_storage[key].extend(new_val)

In [ ]:
len(kv_storage[10])

253

# 2. Добавьте новые значения в key-value хранилище из Ultra-Light  для 1000 случайных ключей (из миллиона). 

## Получение 1000 случайных ключей из 1000000

In [ ]:
keys_random = random.sample(range(1_000_000), 1000)
print("Длина списка случайных значений:", len(keys_random))
print("Первый 10 значений:", keys_random[:10])

Длина списка случайных значений: 1000
Первый 10 значений: [23230, 771756, 131316, 892144, 261913, 375653, 160054, 192078, 488961, 468990]


In [ ]:
print("Длина первого рандомного ключа до добавления:", len(kv_storage[keys_random[0]]))

Длина первого рандомного ключа до добавления: 319


## Добавление

In [ ]:
for key in keys_random:
  num_events = np.random.randint(0,5)
  new_val = np.random.uniform(low=0,high=50,size=num_events).tolist()
  kv_storage[key].extend(new_val)

In [ ]:
print("Длина первого рандомного ключа после добавления:", len(kv_storage[keys_random[0]]))

Длина первого рандомного ключа после добавления: 321


Как видно добавилось **2** значения в список (было 319, стало 321)

# 3. Возьмите любой ключ, по которому были добавлены данные и выведите расчет метрик.

In [ ]:
class KpiBuilder():
  def __init__(self, key):
        self.key = key

  def get(self):
    return kv_storage.get(self.key)

  def mean_20(self):
    return sum(self.get()[:-20])/20.

  def median(self):
    m_ = self.get()
    m_.sort()
    if len(m_)%2==1:
      return m_[(len(m_)//2)]
    else:
      return sum(m_[(len(m_)//2)-1:(len(m_)//2)+1])/2.

  def non_zero_cnt(self):
    v = self.get()
    return len([x for x in v if x > 0])

  def cnt_above_mean(self):
    v = self.get()
    m_ = kv_storage['sum_el']/ float(kv_storage['len_el'])

    print ("Mean across all elements: ",m_)
    return len([x for x in v if x > m_])
  
  ##################################################
  # Свои метрики
  ##################################################

  # 1. Максимальное и минимальное значение
  def min_max(self):
    return (min(self.get()), max(self.get()))
  

  # 2. Разница между суммами значениями выше медианы и ниже медианы
  def number_above_median(self):
    m_ = self.get()
    m_.sort()
    if len(m_)%2==1:
      return sum(m_[(len(m_)//2)+1:]) - sum(m_[:(len(m_)//2)])
    else:
      return sum(m_[(len(m_)//2):]) - sum(m_[:(len(m_)//2)])

  # 3. Среднее значение суммы квадратов
  def mean_sum_squares(self):
    return np.mean([i**2 for i in self.get()])

In [ ]:
# Инициализируем вспомогательное значение sum_el/ len_el суммой элементов
v_len = len([item for sublist in kv_storage.values() for item in sublist])
v_sum = sum([item for sublist in kv_storage.values() for item in sublist])

kv_storage['len_el']=v_len
kv_storage['sum_el']=v_sum

In [ ]:
# Создание экземпляра класса
key_random_9999 = KpiBuilder(keys_random[999])

# **Продемонстрируйте расчет**

## 1. Среднее значение метрик за последние 20 отсчетов

In [ ]:
key_random_9999.mean_20()

56.01324307281626

## 2. Медианное значение метрик за весь период

In [ ]:
key_random_9999.median()

25.680234828836777

## 3. Количество не нулевых метрик

In [ ]:
key_random_9999.non_zero_cnt()

81

## 4. Количество метрик превышающих среднее по всем элементам

In [ ]:
key_random_9999.cnt_above_mean()

Mean across all elements:  25.000332592446107


43

## 5. Максимальное и минимальное значение

In [ ]:
key_random_9999.min_max()

(0.7297068455951516, 48.92319149782298)

## 6. Разница между суммами значениями выше медианы и ниже медианы

In [ ]:
key_random_9999.number_above_median()

1066.6285611890107

## 7. Среднее значение суммы квадратов

In [ ]:
key_random_9999.mean_sum_squares()

841.6413848940276